### ISSUE TITLE
RangeError: Maximum call stack size exceeded

### ISSUE THREAD

**Issue by ephor** [walmartlabs/stylus-relative-loader#12]
When I build the project , I get an error . The project is very large, and apparently the loader can not collect all files and falls with an error. I have 37 entry points. What could it be? 

```
/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:518
        promise._rejectCallback(x.e, false);
                ^

RangeError: Maximum call stack size exceeded
    at Promise._settlePromiseFromHandler /node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:518:17)
    at Promise._settlePromise (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:566:18)
    at Promise._settlePromise0 (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:611:10)
    at Promise._settlePromises (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:690:18)
    at Promise._fulfill (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:635:18)
    at PromiseArray._resolve (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise_array.js:125:19)
    at PromiseArray._promiseFulfilled (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise_array.js:143:14)
    at Promise._settlePromise (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:571:26)
    at Promise._settlePromise0 (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:611:10)
    at Promise._settlePromises (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:690:18)
    at Promise._fulfill (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:635:18)
    at Promise._resolveCallback (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:452:14)
    at Promise._settlePromiseFromHandler (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:521:17)
    at Promise._settlePromise (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:566:18)
    at Promise._settlePromise0 (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:611:10)
    at Promise._settlePromises (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:690:18)
    at Promise._fulfill (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:635:18)
    at PromiseArray._resolve (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise_array.js:125:19)
    at PromiseArray._promiseFulfilled (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise_array.js:143:14)
    at Promise._settlePromise (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:571:26)
    at Promise._settlePromise0 (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:611:10)
    at Promise._settlePromises (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:690:18)
    at Promise._fulfill (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:635:18)
    at Promise._resolveCallback (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:452:14)
    at Promise._settlePromiseFromHandler (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:521:17)
    at Promise._settlePromise (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:566:18)
    at Promise._settlePromise0 (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:611:10)
    at Promise._settlePromises (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:690:18)
    at Promise._fulfill (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:635:18)
    at PromiseArray._resolve (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise_array.js:125:19)
    at PromiseArray._promiseFulfilled (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise_array.js:143:14)
    at Promise._settlePromise (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:571:26)
    at Promise._settlePromise0 (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:611:10)
    at Promise._settlePromises (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:690:18)
    at Promise._fulfill (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:635:18)
    at Promise._resolveCallback (/node_modules/stylus-relative-loader/node_modules/bluebird/js/release/promise.js:452:14)

```

**Comment by exogen** [comment]
That's interesting, I would have guessed the Promise implementation would use some kind of [trampoline](http://raganwald.com/2013/03/28/trampolines-in-javascript.html) strategy so that it wouldn't be possible to blow the stack (just a guess that `.then(tryRender)` within `tryRender` is the culprit). I'll see if switching our Promise library does the trick.

@ephor Is this with an open source project so I can try your build myself?

**Comment by ephor** [comment]
It's not open source project, and unfortunately I do not even know what kind of information I can give you to help sort out this problem. One thing for sure - we have a very large project and a lot of .styl files

**Comment by exogen** [comment]
@ephor Can you try the `bug-promise-call-stack` branch out on your build? You should be able to easily install from GitHub like so:

``` sh
npm install walmartlabs/stylus-relative-loader#bug-promise-call-stack
```

(or change the version of `stylus-relative-loader` in your `package.json` to `walmartlabs/stylus-relative-loader#bug-promise-call-stack`)

**Comment by exogen** [comment]
By the way, how long does it take your build to reach that stack error, once `stylus-relative-loader` starts? I tried reproducing this by making thousands of imports it would need to retry, but the wait time became unbearable attempting only a couple hundred.

**Comment by ephor** [comment]
Thank you so much! It works, and it seems that little bit faster. By the way, we use BEM, and structure as in your example:

```
- ├── a
- │   ├── color.styl
- │   └── index.styl
- ├── b
- │   ├── color.styl
- │   └── index.styl
- ├── color.styl
- └── index.styl
```

Webpack say`s:
Time: 93563ms

Time command say`s:
real    1m33.846s
user    1m35.020s
sys 0m1.108s

It`s a pity that grunt build our project faster.

**Comment by ephor** [comment]
> By the way, how long does it take your build to reach that stack error

Maybe on the middle of build

**Comment by exogen** [comment]
Released in 3.2.1. Thanks @ephor!
